In [1]:
import requests
import pandas as pd
from bs4 import BeautifulSoup as bs
import time

In [2]:
start = time.time()
url = 'https://www.investing.com/stock-screener/Service/SearchStocks'

header = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/79.0.3945.130 Safari/537.36',
    'X-Requested-With': 'XMLHttpRequest'
}

payload = {
    'country[]': '5',  ## US-country code
    'sector': '7,5,12,3,8,9,1,6,2,4,10,11', ## all sectors
    'industry': '81,56,59,41,68,67,88,51,72,47,12,8,50,2,71,9,69,45,46,13,94,102,95,58,100,101,87,31,6,38,79,30,77,'
                '28,5,60,18,26,44,35,53,48,49,55,78,7,86,10,1,34,3,11,62,16,24,20,54,33,83,29,76,37,90,85,82,22,14,'
                '17,19,43,89,96,57,84,93,27,74,97,4,73,36,42,98,65,70,40,99,39,92,75,66,63,21,25,64,61,32,91,52,23,'
                '15,80', ## all industries
    'equityType': 'ORD,DRC,Preferred,Unit,ClosedEnd,REIT,ELKS,OpenEnd,Right,ParticipationShare,CapitalSecurity,'
                  'PerpetualCapitalSecurity,GuaranteeCertificate,IGC,Warrant,SeniorNote,Debenture,ETF,ADR,ETC,ETN',
    'pn': 1,
    'order[col]': 'eq_market_cap',
    'order[dir]': 'd'
}

def run(pages):
    json_list = []
    for i in range(1, pages):
        try:
            payload['pn'] = i
            req_time = time.time()
            req = requests.post(url, data = payload, headers = header).json()
            req_end = time.time()
            print('소요시간 {}초'.format((req_end-req_time)/60))
            print('남은 페이지 수 : {}개'.format(pages - i))   
            json_list += req['hits']
        except Exception as e:
            print(e)
            
    df = pd.DataFrame(json_list)
    return df

pages = int(10914/50) + 1 ## 총 등록된 기업이 10914개이고 페이지 당 50개가 있으니 218페이지를 크롤링해야 합니다.
df = run(pages)

end = time.time()
print('{}분'.format((end-start)/60))

소요시간 0.03306211630503337초
남은 페이지 수 : 218개
소요시간 0.030517876148223877초
남은 페이지 수 : 217개
소요시간 0.037632922331492104초
남은 페이지 수 : 216개
소요시간 0.03033530314763387초
남은 페이지 수 : 215개
소요시간 0.030785266558329263초
남은 페이지 수 : 214개
소요시간 0.03606930573781331초
남은 페이지 수 : 213개
소요시간 0.03460739850997925초
남은 페이지 수 : 212개
소요시간 0.03580445051193237초
남은 페이지 수 : 211개
소요시간 0.031067105134328206초
남은 페이지 수 : 210개
소요시간 0.030041984717051187초
남은 페이지 수 : 209개
소요시간 0.03015298048655192초
남은 페이지 수 : 208개
소요시간 0.03159875472386678초
남은 페이지 수 : 207개
소요시간 0.0349287470181783초
남은 페이지 수 : 206개
소요시간 0.03177058299382528초
남은 페이지 수 : 205개
소요시간 0.03522232373555501초
남은 페이지 수 : 204개
소요시간 0.03567248582839966초
남은 페이지 수 : 203개
소요시간 0.030070261160532633초
남은 페이지 수 : 202개
소요시간 0.02964690128962199초
남은 페이지 수 : 201개
소요시간 0.032230472564697264초
남은 페이지 수 : 200개
소요시간 0.02993663152058919초
남은 페이지 수 : 199개
소요시간 0.03208365043004354초
남은 페이지 수 : 198개
소요시간 0.034368932247161865초
남은 페이지 수 : 197개
소요시간 0.030856200059254966초
남은 페이지 수 : 196개
소요시간 0.03459177414576212초


소요시간 0.03621188004811605초
남은 페이지 수 : 22개
소요시간 0.03653564453125초
남은 페이지 수 : 21개
소요시간 0.0315988302230835초
남은 페이지 수 : 20개
소요시간 0.035921196142832436초
남은 페이지 수 : 19개
소요시간 0.020245277881622316초
남은 페이지 수 : 18개
'hits'
소요시간 0.01893761952718099초
남은 페이지 수 : 17개
'hits'
소요시간 0.019930052757263183초
남은 페이지 수 : 16개
'hits'
소요시간 0.019879627227783202초
남은 페이지 수 : 15개
'hits'
소요시간 0.019847587744394938초
남은 페이지 수 : 14개
'hits'
소요시간 0.020478586355845135초
남은 페이지 수 : 13개
'hits'
소요시간 0.020755537350972495초
남은 페이지 수 : 12개
'hits'
소요시간 0.019963284333546955초
남은 페이지 수 : 11개
'hits'
소요시간 0.02127710183461507초
남은 페이지 수 : 10개
'hits'
소요시간 0.019165412584940592초
남은 페이지 수 : 9개
'hits'
소요시간 0.02114328940709432초
남은 페이지 수 : 8개
'hits'
소요시간 0.02052842378616333초
남은 페이지 수 : 7개
'hits'
소요시간 0.019730091094970703초
남은 페이지 수 : 6개
'hits'
소요시간 0.01938157081604004초
남은 페이지 수 : 5개
'hits'
소요시간 0.019548217455546062초
남은 페이지 수 : 4개
'hits'
소요시간 0.020578217506408692초
남은 페이지 수 : 3개
'hits'
소요시간 0.01894932985305786초
남은 페이지 수 : 2개
'hits'
소요시간 0.0222411195437

In [3]:
df

pair_ID stock_symbol  parent_pair_ID  canonical_to_pair_id  \
0        6408         AAPL               0                     0   
1         252         MSFT               0                     0   
2        6435         AMZN               0                     0   
3        6369        GOOGL               0                     0   
4      100160         GOOG               0                     0   
...       ...          ...             ...                   ...   
9995    52098         MFST               0                     0   
9996    31141        IMRSQ               0                     0   
9997  1136269         XCRP               0                     0   
9998    51449         TMGY               0                     0   
9999    51630         DLOC               0                     0   

      override_country_ID  eq_pe_ratio eq_pe_ratio_eu  eq_market_cap  \
0                       0        25.84          25,84  1420000000000   
1                       0        34.17          34,17  1400000000000   
2                       0        93.84          93,84  1070000000000   
3                       0        30.88          30,88  1040000000000   
4                       0        30.88          30,88  1040000000000   
...                   ...          ...            ...            ...   
9995                    0          NaN            NaN          46330   
9996                    0          NaN            NaN          45810   
9997                    0          NaN            NaN          45460   
9998                    0          NaN            NaN          45410   
9999                    0          NaN            NaN          45330   

     eq_market_cap_eu  eq_one_year_return  ... tech_sum_month_constant  daily  \
0               1,42T               91.47  ...              Strong Buy  -0.65   
1               1,40T               72.80  ...              Strong Buy  -0.54   
2               1,07T               31.87  ...              Strong Buy  -0.57   
3               1,04T               34.68  ...              Strong Buy  -0.39   
4               1,04T               35.39  ...              Strong Buy  -0.33   
...               ...                 ...  ...                     ...    ...   
9995           46,33K              -98.89  ...             Strong Sell   0.00   
9996           45,81K                0.00  ...                     NaN   0.00   
9997           45,46K              -50.00  ...                    Sell   0.00   
9998           45,41K               71.43  ...                    Sell   0.00   
9999           45,33K              -96.67  ...             Strong Sell -33.33   

     daily_eu  daily_frmt   aintcov  aintcov_eu aintcov_us  ttmintcov  \
0       -0,65       -0.65       NaN         NaN        NaN        NaN   
1       -0,54       -0.54       NaN         NaN        NaN        NaN   
2       -0,57       -0.57  18.93359       18,93      18.93   18.95698   
3       -0,39       -0.39       NaN         NaN        NaN        NaN   
4       -0,33       -0.33       NaN         NaN        NaN        NaN   
...       ...         ...       ...         ...        ...        ...   
9995     0,00        0.00  -2.25647       -2,26      -2.26   -0.88306   
9996     0,00        0.00       NaN         NaN        NaN        NaN   
9997     0,00        0.00       NaN         NaN        NaN        NaN   
9998     0,00        0.00       NaN         NaN        NaN        NaN   
9999   -33,33      -33.33  -0.35505       -0,36      -0.36   -2.07450   

     ttmintcov_eu  ttmintcov_us  
0             NaN           NaN  
1             NaN           NaN  
2           18,96         18.96  
3             NaN           NaN  
4             NaN           NaN  
...           ...           ...  
9995        -0,88         -0.88  
9996          NaN           NaN  
9997          NaN           NaN  
9998          NaN           NaN  
9999        -2,07         -2.07  

[10000 rows x 542 columns]

### 10914개 기업이 있다고 알았는데 10000개 밖에 데이터를 주지 않는군요. 이 문제점은 추후에 해결해보도록 하겠습니다. 여기에 각종 재무 데이터는 있습니다만 종목당 일별 주가를 알 수는 없습니다. 그래서 여기의 데이터 중 종목별 ticker와 pair_id를 추출하겠습니다. ticker와 pair_id는 일별 주가를 받아올 때 필요한 정보이기 때문에 추출하는 것입니다.

In [4]:
x = df.copy()
x.set_index('stock_symbol',inplace=True)

In [5]:
x.T

stock_symbol           AAPL   MSFT      AMZN  GOOGL    GOOG     FB    BABA  \
pair_ID                6408    252      6435   6369  100160  26490  941155   
parent_pair_ID            0      0         0      0       0      0       0   
canonical_to_pair_id      0      0         0      0       0      0       0   
override_country_ID       0      0         0      0       0      0       0   
eq_pe_ratio           25.84  34.17     93.84  30.88   30.88  34.63   26.93   
...                     ...    ...       ...    ...     ...    ...     ...   
aintcov_eu              NaN    NaN     18,93    NaN     NaN    NaN     NaN   
aintcov_us              NaN    NaN     18.93    NaN     NaN    NaN     NaN   
ttmintcov               NaN    NaN  18.95698    NaN     NaN    NaN     NaN   
ttmintcov_eu            NaN    NaN     18,96    NaN     NaN    NaN     NaN   
ttmintcov_us            NaN    NaN     18.96    NaN     NaN    NaN     NaN   

stock_symbol           CCI_pa      BRKb      BRKa  ...      NSPX     NUTTQ  \
pair_ID                959747     13834     20562  ...     52631     52315   
parent_pair_ID              0         0         0  ...         0         0   
canonical_to_pair_id        0         0         0  ...         0         0   
override_country_ID         0         0         0  ...         0         0   
eq_pe_ratio             82.25     20.91     20.91  ...         0       NaN   
...                       ...       ...       ...  ...       ...       ...   
aintcov_eu               2,33      1,14      1,14  ...     -0,63     -0,07   
aintcov_us               2.33      1.14      1.14  ...     -0.63     -0.07   
ttmintcov             2.46041  48.45923  48.45923  ...  -0.71909  -0.06612   
ttmintcov_eu             2,46     48,46     48,46  ...     -0,72     -0,07   
ttmintcov_us             2.46     48.46     48.46  ...     -0.72     -0.07   

stock_symbol              CBLRF      CWRL      CRMBQ      MFST  IMRSQ  \
pair_ID                   51516    961113      15812     52098  31141   
parent_pair_ID                0         0          0         0      0   
canonical_to_pair_id          0         0          0         0      0   
override_country_ID           0         0          0         0      0   
eq_pe_ratio                 NaN       NaN        NaN       NaN    NaN   
...                         ...       ...        ...       ...    ...   
aintcov_eu               -31,15     -3,59     -35,49     -2,26    NaN   
aintcov_us               -31.15     -3.59     -35.49     -2.26    NaN   
ttmintcov             -32.44082  -3.08104  -26.38950  -0.88306    NaN   
ttmintcov_eu             -32,44     -3,08     -26,39     -0,88    NaN   
ttmintcov_us             -32.44     -3.08     -26.39     -0.88    NaN   

stock_symbol             XCRP   TMGY      DLOC  
pair_ID               1136269  51449     51630  
parent_pair_ID              0      0         0  
canonical_to_pair_id        0      0         0  
override_country_ID         0      0         0  
eq_pe_ratio               NaN    NaN       NaN  
...                       ...    ...       ...  
aintcov_eu                NaN    NaN     -0,36  
aintcov_us                NaN    NaN     -0.36  
ttmintcov                 NaN    NaN  -2.07450  
ttmintcov_eu              NaN    NaN     -2,07  
ttmintcov_us              NaN    NaN     -2.07  

[541 rows x 10000 columns]

In [6]:
result = x.T.iloc[0]

In [7]:
result

stock_symbol
AAPL        6408
MSFT         252
AMZN        6435
GOOGL       6369
GOOG      100160
          ...   
MFST       52098
IMRSQ      31141
XCRP     1136269
TMGY       51449
DLOC       51630
Name: pair_ID, Length: 10000, dtype: object

In [17]:
# result.to_pickle("../Database/us_pair_id.pickle") 아까 위의 데이터를 저장해두고 싶다면 쓰시면 됩니다.

### 지금부터는 종목당 일별 주가를 가져와 보겠습니다

In [9]:
stock_url = 'https://www.investing.com/instruments/HistoricalDataAjax' 
##payload에 종목 데이터를 넣으면 주가를 보여주는 페이지입니다.

### 우선은 Verizon이라는 한 종목의 주가 데이터를 한 번 가져와보겠습니다.

In [12]:
pair_id = 8127
ticker = 'VZ'
st_date = '01/12/2020'
end_date = '02/12/2020'

payload_stock = {
    'curr_id': pair_id, # 8127. 이 숫자는 investing.com에서 고유하게 종목별로 할당한 id입니다. 그래서 종목당 id를 가져온 것입니다.
    'header': ticker + ' Historical Data', # 'VZ Histrical Data'와 같은 형식으로 ticker를 전달합니다.
    'st_date': st_date, # 01/12/2020 
    'end_date': end_date, # 02/12/2020
    'interval_sec': 'Daily', #일별 데이터입니다.
    'sort_col': 'date', #날짜별 정렬입니다.
    'sort_ord': 'DESC', #내림차순입니다.
    'action': 'historical_data'
}

req = requests.post(stock_url, data=payload_stock, headers=header).text
table = pd.read_html(req)[0]
table

Date  Price   Open   High    Low    Vol. Change %
0   Feb 12, 2020  58.23  58.58  58.86  58.12  20.94M   -0.78%
1   Feb 11, 2020  58.69  60.19  60.35  58.26  22.49M   -2.57%
2   Feb 10, 2020  60.24  59.93  60.24  59.77   8.27M    0.50%
3   Feb 07, 2020  59.94  59.66  60.15  59.54   9.42M    0.84%
4   Feb 06, 2020  59.44  59.20  59.67  59.06  10.28M    0.52%
5   Feb 05, 2020  59.13  58.53  59.19  58.42  10.27M    1.44%
6   Feb 04, 2020  58.29  58.07  58.57  57.65  13.92M    0.41%
7   Feb 03, 2020  58.05  59.61  59.68  58.01  15.57M   -2.34%
8   Jan 31, 2020  59.44  59.34  60.05  59.11  14.28M    0.13%
9   Jan 30, 2020  59.36  58.70  59.39  58.34  16.72M   -0.29%
10  Jan 29, 2020  59.53  60.85  60.92  59.44  13.42M   -1.93%
11  Jan 28, 2020  60.70  59.98  61.00  59.87  15.37M    1.32%
12  Jan 27, 2020  59.91  59.84  60.36  59.62  11.48M   -0.61%
13  Jan 24, 2020  60.28  60.50  60.66  60.11  10.18M   -0.38%
14  Jan 23, 2020  60.51  60.38  60.53  60.00  11.11M    0.05%
15  Jan 22, 2020  60.48  60.45  60.82  60.28   8.85M    0.27%
16  Jan 21, 2020  60.32  60.20  60.47  60.10  12.86M    0.32%
17  Jan 17, 2020  60.13  59.69  60.15  59.64  15.62M    0.77%
18  Jan 16, 2020  59.67  59.38  59.85  59.22  10.69M    0.96%
19  Jan 15, 2020  59.10  59.25  59.42  58.87  10.29M    0.10%
20  Jan 14, 2020  59.04  59.00  59.13  58.76  12.70M   -0.19%
21  Jan 13, 2020  59.15  59.02  59.37  58.78  14.16M    0.27%

### 위의 코드를 바탕으로 10000개 종목의 일별 데이터를 가져올 수 있습니다. 그런데 시간이 너무 오래걸려서 31개 종목만 돌려보겠습니다. 31개를 돌리는데 37.96초가 나오므로 10000개라면 204분, 즉 3시간이 넘게 걸리네요... 1달 데이터만 가져오긴 하는 거지만 상당히 느리긴합니다....

In [15]:
st = time.time()

tickers = list(result.index)
st_date = '01/12/2020'
end_date = '02/12/2020'

payload_stock = {
    'header': ticker + ' Historical Data', # 'VZ Histrical Data'와 같은 형식으로 ticker를 전달합니다.
    'st_date': st_date, # 01/12/2020 
    'end_date': end_date, # 02/12/2020
    'interval_sec': 'Daily', #일별 데이터입니다.
    'sort_col': 'date', #날짜별 정렬입니다.
    'sort_ord': 'DESC', #내림차순입니다.
    'action': 'historical_data'
}

daily_prices = []
for idx,ticker in enumerate(tickers):
    pair_id = result.loc[ticker]
    payload_stock['curr_id'] = pair_id
    payload_stock['header'] = ticker + ' Historical Data'
    print('{} 데이터 수집 중...'.format(ticker))
    req = requests.post(stock_url, data=payload_stock, headers=header).text
    soup = bs(req, 'lxml').find('table')
    daily_prices.append(str(soup))
    if idx==30:
        break

daily_price_stocks = " ".join(daily_prices)
daily_price_stocks_df = pd.read_html(daily_price_stocks)
end = time.time()
print(end-st)

AAPL 데이터 수집 중...
MSFT 데이터 수집 중...
AMZN 데이터 수집 중...
GOOGL 데이터 수집 중...
GOOG 데이터 수집 중...
FB 데이터 수집 중...
BABA 데이터 수집 중...
CCI_pa 데이터 수집 중...
BRKb 데이터 수집 중...
BRKa 데이터 수집 중...
TCTZF 데이터 수집 중...
TCEHY 데이터 수집 중...
V 데이터 수집 중...
JPM_pa 데이터 수집 중...
JPM_pf 데이터 수집 중...
JPM_pg 데이터 수집 중...
JPM_ph 데이터 수집 중...
JPM 데이터 수집 중...
JPM_pe 데이터 수집 중...
JPM_pc 데이터 수집 중...
JNJ 데이터 수집 중...
SSNLF 데이터 수집 중...
MA 데이터 수집 중...
WMT 데이터 수집 중...
BAC_pe 데이터 수집 중...
BAC_pl 데이터 수집 중...
BAC_py 데이터 수집 중...
BML_pl 데이터 수집 중...
BAC 데이터 수집 중...
BAC_pw 데이터 수집 중...
BML_pg 데이터 수집 중...
37.9678258895874


In [16]:
daily_price_stocks_df[0] 
## daily_price_stocks_df는 list로, 요소마다 아까 추출한 티커의 1달치 일별 데이터가 있습니다.
## 이 데이터만 보고는 어떤 종목의 데이터인지 모르므로, 티커를 추가해주겠습니다.

Date   Price    Open    High     Low    Vol. Change %
0   Feb 12, 2020  327.20  321.47  327.22  321.47  28.43M    2.37%
1   Feb 11, 2020  319.61  323.60  323.90  318.71  23.58M   -0.60%
2   Feb 10, 2020  321.55  314.18  321.55  313.85  27.34M    0.47%
3   Feb 07, 2020  320.03  322.37  323.40  318.00  29.42M   -1.59%
4   Feb 06, 2020  325.21  322.57  325.22  320.26  26.36M    1.17%
5   Feb 05, 2020  321.45  323.52  324.76  318.95  29.71M    0.82%
6   Feb 04, 2020  318.85  315.31  319.64  313.63  34.15M    3.30%
7   Feb 03, 2020  308.66  304.30  313.49  302.22  43.50M   -0.27%
8   Jan 31, 2020  309.51  320.93  322.68  308.29  49.90M   -4.43%
9   Jan 30, 2020  323.87  320.54  324.09  318.75  31.69M   -0.14%
10  Jan 29, 2020  324.34  324.45  327.85  321.38  54.15M    2.09%
11  Jan 28, 2020  317.69  312.60  318.40  312.19  40.56M    2.83%
12  Jan 27, 2020  308.95  310.06  311.77  304.88  40.49M   -2.94%
13  Jan 24, 2020  318.31  320.25  323.33  317.52  36.63M   -0.29%
14  Jan 23, 2020  319.23  317.92  319.56  315.65  26.12M    0.48%
15  Jan 22, 2020  317.70  318.58  319.99  317.31  25.46M    0.36%
16  Jan 21, 2020  316.57  317.19  319.02  316.00  27.71M   -0.68%
17  Jan 17, 2020  318.73  316.27  318.74  315.00  34.45M    1.11%
18  Jan 16, 2020  315.24  313.59  315.70  312.09  27.21M    1.25%
19  Jan 15, 2020  311.34  311.85  315.50  309.55  30.48M   -0.43%
20  Jan 14, 2020  312.68  316.70  317.57  312.17  40.65M   -1.35%
21  Jan 13, 2020  316.96  311.64  317.07  311.15  30.52M    2.14%

In [18]:
result.head()

stock_symbol
AAPL       6408
MSFT        252
AMZN       6435
GOOGL      6369
GOOG     100160
Name: pair_ID, dtype: object

In [25]:
for idx, stock in enumerate(daily_price_stocks_df):
    stock['ticker'] = result.index[idx]
    daily_price_stocks_df[idx] = stock   

In [27]:
daily_price_stocks_df[0]

Date   Price    Open    High     Low    Vol. Change % ticker
0   Feb 12, 2020  327.20  321.47  327.22  321.47  28.43M    2.37%   AAPL
1   Feb 11, 2020  319.61  323.60  323.90  318.71  23.58M   -0.60%   AAPL
2   Feb 10, 2020  321.55  314.18  321.55  313.85  27.34M    0.47%   AAPL
3   Feb 07, 2020  320.03  322.37  323.40  318.00  29.42M   -1.59%   AAPL
4   Feb 06, 2020  325.21  322.57  325.22  320.26  26.36M    1.17%   AAPL
5   Feb 05, 2020  321.45  323.52  324.76  318.95  29.71M    0.82%   AAPL
6   Feb 04, 2020  318.85  315.31  319.64  313.63  34.15M    3.30%   AAPL
7   Feb 03, 2020  308.66  304.30  313.49  302.22  43.50M   -0.27%   AAPL
8   Jan 31, 2020  309.51  320.93  322.68  308.29  49.90M   -4.43%   AAPL
9   Jan 30, 2020  323.87  320.54  324.09  318.75  31.69M   -0.14%   AAPL
10  Jan 29, 2020  324.34  324.45  327.85  321.38  54.15M    2.09%   AAPL
11  Jan 28, 2020  317.69  312.60  318.40  312.19  40.56M    2.83%   AAPL
12  Jan 27, 2020  308.95  310.06  311.77  304.88  40.49M   -2.94%   AAPL
13  Jan 24, 2020  318.31  320.25  323.33  317.52  36.63M   -0.29%   AAPL
14  Jan 23, 2020  319.23  317.92  319.56  315.65  26.12M    0.48%   AAPL
15  Jan 22, 2020  317.70  318.58  319.99  317.31  25.46M    0.36%   AAPL
16  Jan 21, 2020  316.57  317.19  319.02  316.00  27.71M   -0.68%   AAPL
17  Jan 17, 2020  318.73  316.27  318.74  315.00  34.45M    1.11%   AAPL
18  Jan 16, 2020  315.24  313.59  315.70  312.09  27.21M    1.25%   AAPL
19  Jan 15, 2020  311.34  311.85  315.50  309.55  30.48M   -0.43%   AAPL
20  Jan 14, 2020  312.68  316.70  317.57  312.17  40.65M   -1.35%   AAPL
21  Jan 13, 2020  316.96  311.64  317.07  311.15  30.52M    2.14%   AAPL

이제 티커가 추가 되었네요! 이제 데이터를 병합하여 저장을 하겠습니다.

In [28]:
save = pd.concat(daily_price_stocks_df)
save.to_pickle('./us_daily_one_month.pickle') # ./는 상대경로 중 현재 폴더에 저장하는 것입니다. ../는 상위폴더입니다.
# pickle로 저장하는 이유
# 읽고 쓰는 시간이 짧습니다.
# 데이터형이 보존됩니다. csv는 '001'이라는 string이 1이라는 int로 변환됩니다.

In [29]:
save

Date   Price    Open    High     Low    Vol. Change %  ticker
0   Feb 12, 2020   327.2  321.47  327.22  321.47  28.43M    2.37%    AAPL
1   Feb 11, 2020  319.61   323.6   323.9  318.71  23.58M   -0.60%    AAPL
2   Feb 10, 2020  321.55  314.18  321.55  313.85  27.34M    0.47%    AAPL
3   Feb 07, 2020  320.03  322.37   323.4     318  29.42M   -1.59%    AAPL
4   Feb 06, 2020  325.21  322.57  325.22  320.26  26.36M    1.17%    AAPL
..           ...     ...     ...     ...     ...     ...      ...     ...
16  Jan 17, 2020   22.05   22.19   22.19   21.82   1.48K    1.29%  BML_pg
17  Jan 16, 2020   21.77   21.86   22.19   21.68   6.41K   -1.67%  BML_pg
18  Jan 15, 2020   22.14    22.1   22.14    22.1   1.02K    0.27%  BML_pg
19  Jan 14, 2020   22.08    21.9   22.08    21.9   1.43K    0.59%  BML_pg
20  Jan 13, 2020   21.95      22   22.06   21.69   3.54K   -0.50%  BML_pg

[582 rows x 8 columns]

## 지금부터는 investing.com의 실시간 데이터를 가져와보겠습니다. investing.com은 5분 단위로 업데이트합니다. yahoo finance 또한 15분 지연 데이터를 주기 때문에 investing.com을 사용했습니다.

In [30]:
from datetime import datetime

In [33]:
import json

### 아래의 url은 현재가를 받아오는 url입니다.

In [41]:
url = 'https://advcharts.investing.com/advinion2016/advanced-charts/1/1/8/GetRecentHistory'
payload = {
    'strSymbol': 8127, 
    ## 이 symbol은 verizon의 pair_id입니다. 
    'iTop': '1500',
    'strPriceType': 'bid',
    'strFieldsMode': 'allFields',
    'strExtraData': 'lang_ID=1',
    'strTimeFrame': '5M' ## 5분 단위로 받아오겠습니다. 5분이 최소단위입니다.
}

header = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36'
}

In [42]:
iter_=0
while True:
    if iter_==0: ## 처음 시작할 떄는 00초의 정각이 될 때까지 기다리게 하겠습니다. 
        wait_time = 60 - datetime.now().second
        print('현재 시간 {0}, 대기시간 {1}초'.format(datetime.now().strftime('%H:%M:%S'), wait_time))
        time.sleep(wait_time)
        iter_+=1
    
    else:        
        req = requests.get(url, params=payload, headers=header).text
        latest = json.loads(req)['data'][-1]
        latest_df = pd.DataFrame([latest])
        now_minute = datetime.now().minute
        print(latest_df[['date','open','close', 'low', 'high', 'volume', 'volume_ac']])
        time.sleep(60) ## 1분 단위로 출력해보겠습니다.

현재 시간 15:44:11, 대기시간 49초
                  date       open      close        low       high  volume  \
0  2020-02-13 23:30:00  58.650002  58.650002  58.650002  58.650002   12724   

   volume_ac  
0   19460059  
                  date       open      close        low       high  volume  \
0  2020-02-13 23:30:00  58.650002  58.650002  58.650002  58.650002   12724   

   volume_ac  
0   19460059  
                  date       open      close        low       high  volume  \
0  2020-02-13 23:30:00  58.650002  58.650002  58.650002  58.650002   12724   

   volume_ac  
0   19460059  
                  date       open      close        low       high  volume  \
0  2020-02-13 23:30:00  58.650002  58.650002  58.650002  58.650002   12724   

   volume_ac  
0   19460059  
                  date       open      close        low       high  volume  \
0  2020-02-13 23:30:00  58.650002  58.650002  58.650002  58.650002   12724   

   volume_ac  
0   19460059  


KeyboardInterrupt: 

### 제 컴퓨터 시간은 14일 15시 40분인데 date에 뜨는 시간이 어제 23시 30분이네요.  시차 때문에 그런 것 같습니다. 미 증시 개장 시간이 아니라 다시 해봐야겠군요.

In [49]:
stock_url = 'https://www.investing.com/instruments/HistoricalDataAjax' 
pair_id = 940767
ticker = 'CFRX'
st_date = '01/12/2020'
end_date = '02/12/2020'

header = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/79.0.3945.130 Safari/537.36',
    'X-Requested-With': 'XMLHttpRequest'
}

payload_stock = {
    'curr_id': pair_id, # 8127. 이 숫자는 investing.com에서 고유하게 종목별로 할당한 id입니다. 그래서 종목당 id를 가져온 것입니다.
    'header': ticker + ' Historical Data', # 'VZ Histrical Data'와 같은 형식으로 ticker를 전달합니다.
    'st_date': st_date, # 01/12/2020 
    'end_date': end_date, # 02/12/2020
    'interval_sec': 'Daily', #일별 데이터입니다.
    'sort_col': 'date', #날짜별 정렬입니다.
    'sort_ord': 'DESC', #내림차순입니다.
    'action': 'historical_data'
}

req = requests.post(stock_url, data=payload_stock, headers=header).text
table = pd.read_html(req)[0]
table

Date   Price   Open    High     Low     Vol. Change %
0   Feb 12, 2020   7.570   7.51   7.810   7.510  372.25K    1.07%
1   Feb 11, 2020   7.490   7.74   7.870   7.300  145.82K   -3.23%
2   Feb 10, 2020   7.740   6.91   8.150   6.910  525.04K   13.82%
3   Feb 07, 2020   6.800   7.19   7.190   6.660  177.38K   -5.42%
4   Feb 06, 2020   7.190   7.47   7.700   7.023  159.38K   -4.01%
5   Feb 05, 2020   7.490   7.57   8.000   7.269  223.15K   -3.97%
6   Feb 04, 2020   7.800   6.50   8.450   5.906  477.87K    6.85%
7   Feb 03, 2020   7.300   8.20   8.209   7.200  151.33K  -12.05%
8   Jan 31, 2020   8.300   8.40   8.900   8.100   87.01K   -4.44%
9   Jan 30, 2020   8.686   9.05   9.050   8.001  143.48K   -5.08%
10  Jan 29, 2020   9.151   9.20  10.400   9.020   63.81K   -0.52%
11  Jan 28, 2020   9.199  10.40  10.400   8.500  201.23K   -8.01%
12  Jan 27, 2020  10.000  10.70  10.800   9.800  129.29K   -9.09%
13  Jan 24, 2020  11.000  11.30  11.500  10.600  121.85K   -1.79%
14  Jan 23, 2020  11.200  11.50  11.900  11.100  277.22K   -1.32%
15  Jan 22, 2020  11.350  11.40  11.500  10.600  301.32K    1.34%
16  Jan 21, 2020  11.200  10.80  11.500  10.700  168.72K    7.69%
17  Jan 17, 2020  10.400  10.50  10.800  10.101  206.58K   -6.31%
18  Jan 16, 2020  11.100  11.60  11.697  10.700  164.25K   -5.93%
19  Jan 15, 2020  11.800  12.40  12.496  11.100  261.65K   -4.07%
20  Jan 14, 2020  12.300  12.90  13.000  11.100  352.06K    3.36%
21  Jan 13, 2020  11.900  11.80  13.399  11.000  678.90K   13.33%